src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -846,143 +846,160 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
+                                            (format!("calls in {}s are limited to \
                                             tuple structs and tuple variants",
                                             self.mode),
-                                    Some("a limited form of compile-time function \
-                                        evaluation is available on a nightly \
+                                            Some("a limited form of compile-time function \
+                                                 evaluation is available on a nightly \
                                         compiler via `const fn`"))
-                                } else {
-                                    (format!("calls in {}s are limited \
+                                        } else {
+                                            (format!("calls in {}s are limited \
                                             to constant functions, \
                                             tuple structs and tuple variants",
                                             self.mode),
-                                    None)
-                                };
-                                let mut err = struct_span_err!(
-                                    self.tcx.sess,
-                                    self.span,
-                                    E0015,
-                                    "{}",
-                                    msg,
-                                );
-                                if let Some(note) = note {
-                                    err.span_note(self.span, note);
+                                            None)
+                                        };
+                                    let mut err = struct_span_err!(
+                                        self.tcx.sess,
+                                        self.span,
+                                        E0015,
+                                        "{}",
+                                        msg,
+                                        );
+                                    if let Some(note) = note {
+                                        err.span_note(self.span, note);
+                                    }
+                                    err.emit();
                                 }
-                                err.emit();
                             }
                         }
                     }
+                },
+                ty::FnPtr(_) => {
+                    if self.mode != Mode::Fn {
+                        let mut err = self.tcx.sess.struct_span_err(
+                            self.span,
+                            &format!("function pointers are not allowed in const fn"));
+                        err.emit();
+                    }
+                },
+                _ => {
+                    span_bug!(self.span, "this is not allowed in const fn");
```

Comment:
I think it would be strictly less ICEy if we just called `not_const` here and early returned